File: haskell/problem4.hs

-- A palindromic number reads the same both ways. The largest palindrome made
--  from the product of two 2-digit numbers is 9009 = 91 × 99.
--
-- Find the largest palindrome made from the product of two 3-digit numbers.

toList :: Integer -> [Integer]
toList 0 = []
toList n = n `mod` 10 : toList (n `div` 10)

isPalindrome :: Integer -> Bool
isPalindrome n = toList n == reverse (toList n)

largetsPalindrome :: Integer
largetsPalindrome = maximum [p | x <- [100..999], y <- [x..999], let p = x * y, isPalindrome p]

File: js/problems/Problem4.js
/*
A palindromic number reads the same both ways. The largest palindrome made
 from the product of two 2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit numbers.
*/

const isPalidrom = (number) => {
  let numStr = number.toString();

  return numStr === numStr.split('').reverse().join('');
};

const largestPalindrome = (n = 3) => {
  if (n > 4) return console.warn('Well depending on your PC this will take between a minute and infinity /think infinity/, so NO!');

  let max = 0;
  let low = Math.pow(10, n - 1);
  let hi = Math.pow(10, n);

  for (var i = low; i < hi; i++) {
    for (var j = i; j < hi; j++) {
      let p = i * j;
      if (isPalidrom(p) && p > max) {
        max = p;
      }
    }
  }

  return max;
};

module.exports = largestPalindrome;
